src/Fields/field.jl

```diff
@@ -63,66 +64,149 @@ Construct a `Field` on some architecture `arch` and a `grid` with some `data`.
+              data=zeros(FT, arch, grid, (Cell, Cell, Face))
 
 Return a `Field{Cell, Cell, Face}` on architecture `arch` and `grid` containing `data`
 with field boundary conditions `bcs`.
 """
-ZFaceField(arch::AbstractArchitecture, grid, bcs=WVelocityBoundaryConditions(grid), data=zeros(arch, grid)) =
-    Field(Cell, Cell, Face, arch, grid, bcs, data)
+function ZFaceField(arch::AbstractArchitecture, grid, 
+                    bcs=WVelocityBoundaryConditions(grid), 
+                    data=zeros(eltype(grid), arch, grid, (Cell, Cell, Face))) 
 
-CellField(FT::DataType, arch, grid, bcs=TracerBoundaryConditions(grid)) =
-    CellField(arch, grid, bcs, zeros(FT, arch, grid))
+    return Field(Cell, Cell, Face, arch, grid, bcs, data)
+end
+
+CellField(FT::DataType, arch, grid, bcs=TracerBoundaryConditions(grid)) = 
+    CellField(arch, grid, bcs, zeros(FT, arch, grid, (Cell, Cell, Cell)))
 
 XFaceField(FT::DataType, arch, grid, bcs=UVelocityBoundaryConditions(grid)) =
-    XFaceField(arch, grid, bcs, zeros(FT, arch, grid))
+    XFaceField(arch, grid, bcs, zeros(FT, arch, grid, (Face, Cell, Cell)))
 
 YFaceField(FT::DataType, arch, grid, bcs=VVelocityBoundaryConditions(grid)) =
-    YFaceField(arch, grid, bcs, zeros(FT, arch, grid))
+    YFaceField(arch, grid, bcs, zeros(FT, arch, grid, (Cell, Face, Cell)))
 
 ZFaceField(FT::DataType, arch, grid, bcs=WVelocityBoundaryConditions(grid)) =
-    ZFaceField(arch, grid, bcs, zeros(FT, arch, grid))
+    ZFaceField(arch, grid, bcs, zeros(FT, arch, grid, (Cell, Cell, Face)))
+
+#####
+##### Functions for querying fields
+#####
 
 location(a) = nothing
-location(::AbstractField{X, Y, Z}) where {X, Y, Z} = (X, Y, Z)
 
-architecture(f::Field)       = architecture(f.data)
+"Returns the location `(X, Y, Z)` of an `AbstractField{X, Y, Z}`."
+location(::AbstractField{X, Y, Z}) where {X, Y, Z} = (X, Y, Z) # note no instantiation
+
+x_location(::AbstractField{X}) where X               = X
+y_location(::AbstractField{X, Y}) where {X, Y}       = Y
+z_location(::AbstractField{X, Y, Z}) where {X, Y, Z} = Z
+
+"Returns the architecture where the field data `f.data` is stored."
+architecture(f::Field) = architecture(f.data)
 architecture(o::OffsetArray) = architecture(o.parent)
 
-@inline size(f::AbstractField) = size(f.grid)
+"Returns the length of a field's `data`."    
 @inline length(f::Field) = length(f.data)
 
+
+"Returns the topology of a fields' `grid`."
+@inline topology(f) = topology(f.grid)
+@inline x_topology(f) = x_topology(f.grid)
+@inline y_topology(f) = y_topology(f.grid)
+@inline z_topology(f) = z_topology(f.grid)
+
+"""
+Returns the length of a field located at `Cell` centers along a grid 
+dimension of length `N` and with halo points `H`.
+"""
+length(loc, topo, N, H=0) = N + 2H
+
+"""
+Returns the length of a field located at cell `Face`s along a grid 
+dimension of length `N` and with halo points `H`.
+"""
+length(::Type{Face}, ::Bounded, N, H=0) = N + 1 + 2H
```

Comment:
Hmmm, can we reasonably assume `H=0` by default? The grid always specified the size of the halo regions so why not explicitly pass `H`?

Reply:
I wonder if it's worth picking another name besides `length` as it could be confused with `Base.length` which takes a single argument (don't think you're trying to overload `Base.length`).

Maybe `dimension_length`?

Reply:
I really did intend to overload `length`, though I think you're right that maybe a different name would be better here. The function `size` does take a dimension argument, but I can't make the argument pattern fit into that box.

Reply:
The default `H=0` is not an assumption, but a convenience method.